com.unity.netcode.gameobjects/Runtime/Serialization/BufferSerializer.cs

```diff
@@ -0,0 +1,312 @@
+using System;
+using UnityEngine;
+
+namespace Unity.Netcode
+{
+    /// <summary>
+    /// Two-way serializer wrapping FastBufferReader or FastBufferWriter.
+    /// 
+    /// Implemented as a ref struct for two reasons:
+    /// 1. The BufferSerializer cannot outlive the FBR/FBW it wraps or using it will cause a crash
+    /// 2. The BufferSerializer must always be passed by reference and can't be copied
+    ///
+    /// Ref structs help enforce both of those rules: they can't out live the stack context in which they were
+    /// created, and they're always passed by reference no matter what.
+    ///
+    /// BufferSerializer doesn't wrapp FastBufferReader or FastBufferWriter directly because it can't.
+    /// ref structs can't implement interfaces, and in order to be able to have two different implementations with
+    /// the same interface (which allows us to avoid an "if(IsReader)" on every call), the thing directly wrapping
+    /// the struct has to implement an interface. So IBufferSerializerImplementation exists as the interface,
+    /// which is implemented by a normal struct, while the ref struct wraps the normal one to enforce the two above
+    /// requirements. (Allowing direct access to the IBufferSerializerImplementation struct would allow dangerous
+    /// things to happen because the struct's lifetime could outlive the Reader/Writer's.)
+    /// </summary>
+    /// <typeparam name="TImplementation">The implementation struct</typeparam>
+    public ref struct BufferSerializer<TImplementation> where TImplementation : IBufferSerializerImplementation
```

Comment:
I see a similar conversation initiated by Luke above but I'd like to kick-off another one.
Can we compare `if (IsReading)` approach vs interface then reader or writer impl behind the scenes approach to actually see performance difference?
If it's not that different, I'd suggest keeping things simple, less abstract and in one place (hence the existing `NetworkSerializer` implementation).

Reply:
Branching is pretty expensive. IIRC when I compared it before it was like a 2-3x cost increase to branch. I'll rerun that test in the morning since it's been a while.

Reply:
Responded to this in the above comment.